### DISCUSSION TITLE
Support for accidentals like E#?

### DISCUSSION THREAD

**Discussion by nimblespark** [CoderLine/alphaTab#1717] [Ideas]
I'm having a bit of a problem when writing sheet music using alphaTex. It seems like whenever I try to do an accidental like E# or Cb, it just replaces it with a C4 (middle c). You might think that I could just use the enharmonic equivalent instead (for example F instead of E#). However, from a music theory standpoint, it is often the case that I really need to write E# and F just doesn't work. Without support for this, I would have to switch to something like abc.js to represent sheet music. I much prefer this library though so I don't really want to do that haha.

So I'm wondering if support could be added for these types of accidentals.
Specifically:
Cb, Fb, E#, B#, (most important)

Also:
A##, B##, C##, D##, E##, F##, G##, 
Abb, Bbb, Cbb, Dbb, Ebb, Fbb, Gbb

**Comment by Danielku15** [comment]
Can you provide some concrete examples with actual and expected behaviors? 

I realized that the docs around the non-fretted alphaTex are a bit sparse. It currently works like this: 

You express the "note values" in one of these ways:

* as `Fret.String` and the height is determined based on tuning of the track
* as articulation names for percussion tracks (special)
* `x` for dead notes
* `-` for tied notes
* "Tuning" expressed in "<Note><Octave>" where the note is the index within the octave: 
   * c=0
   * c#/db=1
   * d=2
   * d#/eb=3
   * e=4
   * f=5
   * f#/gb=6
   * g=7
   * g#/ab=8
   * a=9
   * a#/bb=10
   * b=11

The display of the note is then a different topic where during engraving it is determined what accidentals and line it gets (respecting things like key signatures and clefs). 

As of today writing F#9 doesn't give you a `#` accidetnal and Gb9 a `b` accidental. Similarly there is now way to specify advanced accidentals like a double-flat or double-sharp.

alphaTab has support for these accidentals (e.g. when using guitar pro they are respected), its just that alphaTex doesn't allow you to specify it:
![image](https://github.com/user-attachments/assets/65e2b56a-ea32-42bc-948b-2935f671bf6c)

Do I understand you correctly that you'd like to be able to express the above notation in a alphaTex like this?
```
G#4 G4 G##4 G4 | Gb4 G4 Gbb4 G4
// or
G#4 G4 Gx4 G4 | Gb4 G4 Gbb4 G4
```

If yes I can take it up as a work item to add support for this. Just want to be sure I do the right thing 😉 


Side note: Key Signatures have to be accounted, a G#4 does not guarantee you a `#` accidental on the note as the key signature might define it already.
![image](https://github.com/user-attachments/assets/b751fa7c-eab4-4dde-8d74-6166f8efa7a6)

This would be: 
```
B#4 B4 B##4 B#4 | Bb4 B4 Bbb4 B#4 
```

**Reply by nimblespark** [reply]
> Do I understand you correctly that you'd like to be able to express the above notation in a alphaTex like this?

```
G#4 G4 G##4 G4 | Gb4 G4 Gbb4 G4
// or
G#4 G4 Gx4 G4 | Gb4 G4 Gbb4 G4
```

Yes, that's correct. I would prefer using ## to refer to double sharps but x works too.

> Side note: Key Signatures have to be accounted, a G#4 does not guarantee you a # accidental on the note as the key signature might define it already.

And yeah, this is true. If the accidental is in the key signature, there's no need to show it again.

> Can you provide some concrete examples with actual and expected behaviors?

My expected behavior is exactly as you described.
```
B#4 B4 B##4 B#4 | Bb4 B4 Bbb4 B#4
```
should make
![image](https://github.com/user-attachments/assets/90e527b5-0f85-4b41-918a-3c0c92b8f64c)
Also, something like 
```
A4 A#4 A#4 A4
```
should write this:
<img width="218" alt="Screenshot 2024-10-25 at 8 46 08 AM" src="https://github.com/user-attachments/assets/97fcf079-cdff-452e-b530-fe88cdec3f2e">
A note that already has an accidental applied earlier in the bar doesn't need it again. 

Furthermore, when writing guitar tab with notes out of the key, there should probably be some way to specify which note to display on the sheet music. 
For example, something like `5.1#` could indicate that you want this note to be represented in sheet music as an A# and `5.1b` could indicate that you want it to be represented as a Bb. There would be a default (it might use Bb by default) but you should be able to override the default if you want the note expressed differently. This isn't a high priority for me though and you might consider it a totally separate issue.

Specifically, the reason why I want to be able to represent notes with these more complicated accidentals, is I have a bunch of code that produces these accidentals and I would like to be able to display them on alphaTab.  

Thank you for your detailed response. 
Just let me know what you think the timeline on this will be.

**Reply by Danielku15** [reply]
Added https://github.com/CoderLine/alphaTab/issues/1719
